File: src/features/export/utils/maskToPolygon.ts
/**
 * Convert a binary mask to polygon using Moore-Neighbor tracing algorithm
 *
 * Moore-Neighbor Tracing is a contour tracing algorithm that follows the boundary
 * of a binary region by examining the 8-connected neighbors in a specific order.
 */

export interface Point {
  x: number;
  y: number;
}

/**
 * Extract contour from a binary mask (canvas)
 * @param maskCanvas - Canvas element containing the mask
 * @returns Array of polygon points
 */
export function maskToPolygon(maskCanvas: HTMLCanvasElement): Point[] {
  const ctx = maskCanvas.getContext('2d');
  if (!ctx) return [];

  const width = maskCanvas.width;
  const height = maskCanvas.height;

  // Get image data
  const imageData = ctx.getImageData(0, 0, width, height);
  const data = imageData.data;

  // Create binary mask (1 = white pixel, 0 = transparent/black)
  const mask: number[][] = [];
  for (let y = 0; y < height; y++) {
    mask[y] = [];
    for (let x = 0; x < width; x++) {
      const idx = (y * width + x) * 4;
      const alpha = data[idx + 3];
      // Consider pixel as part of mask if alpha > 128
      mask[y][x] = alpha > 128 ? 1 : 0;
    }
  }

  // Find starting point (first white pixel from top-left)
  let startX = -1;
  let startY = -1;

  outerLoop: for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      if (mask[y][x] === 1) {
        startX = x;
        startY = y;
        break outerLoop;
      }
    }
  }

  // No white pixels found
  if (startX === -1 || startY === -1) {
    return [];
  }

  // Moore-Neighbor tracing
  const contour = mooreNeighborTrace(mask, startX, startY, width, height);

  return contour;
}

/**
 * Moore-Neighbor tracing algorithm
 */
function mooreNeighborTrace(
  mask: number[][],
  startX: number,
  startY: number,
  width: number,
  height: number
): Point[] {
  const contour: Point[] = [];

  // 8-connected neighbors (clockwise from top-left)
  // Directions: 0=N, 1=NE, 2=E, 3=SE, 4=S, 5=SW, 6=W, 7=NW
  const dx = [0, 1, 1, 1, 0, -1, -1, -1];
  const dy = [-1, -1, 0, 1, 1, 1, 0, -1];

  let x = startX;
  let y = startY;
  let direction = 7; // Start looking from NW (top-left)

  const maxIterations = width * height * 2; // Safety limit
  let iterations = 0;

  do {
    contour.push({ x, y });

    // Search for next boundary pixel
    let found = false;
    for (let i = 0; i < 8; i++) {
      const checkDir = (direction + i) % 8;
      const nx = x + dx[checkDir];
      const ny = y + dy[checkDir];

      // Check bounds
      if (nx >= 0 && nx < width && ny >= 0 && ny < height) {
        if (mask[ny][nx] === 1) {
          // Found next boundary pixel
          x = nx;
          y = ny;
          // Update search direction (backtrack 2 steps for next iteration)
          direction = (checkDir + 6) % 8; // -2 in circular array
          found = true;
          break;
        }
      }
    }

    if (!found) {
      // No neighbor found, might be isolated pixel
      break;
    }

    iterations++;
    if (iterations > maxIterations) {
      console.warn('Moore-Neighbor tracing: max iterations reached');
      break;
    }
  } while (x !== startX || y !== startY || contour.length < 2);

  // Remove duplicate start/end point
  if (contour.length > 1 && contour[0].x === contour[contour.length - 1].x && contour[0].y === contour[contour.length - 1].y) {
    contour.pop();
  }

  return contour;
}

/**
 * Extract multiple contours from a mask (handles multiple disconnected regions)
 */
export function maskToPolygons(maskCanvas: HTMLCanvasElement, minArea: number = 10): Point[][] {
  const ctx = maskCanvas.getContext('2d');
  if (!ctx) return [];

  const width = maskCanvas.width;
  const height = maskCanvas.height;

  // Get image data
  const imageData = ctx.getImageData(0, 0, width, height);
  const data = imageData.data;

  // Create binary mask
  const mask: number[][] = [];
  for (let y = 0; y < height; y++) {
    mask[y] = [];
    for (let x = 0; x < width; x++) {
      const idx = (y * width + x) * 4;
      const alpha = data[idx + 3];
      mask[y][x] = alpha > 128 ? 1 : 0;
    }
  }

  const polygons: Point[][] = [];
  const visited: boolean[][] = Array(height)
    .fill(0)
    .map(() => Array(width).fill(false));

  // Find all contours
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      if (mask[y][x] === 1 && !visited[y][x]) {
        const contour = mooreNeighborTrace(mask, x, y, width, height);

        // Mark visited pixels
        for (const point of contour) {
          if (point.y >= 0 && point.y < height && point.x >= 0 && point.x < width) {
            visited[point.y][point.x] = true;
          }
        }

        // Filter by minimum area
        if (contour.length >= 3) {
          const area = calculatePolygonArea(contour);
          if (area >= minArea) {
            polygons.push(contour);
          }
        }
      }
    }
  }

  return polygons;
}

/**
 * Calculate polygon area using shoelace formula
 */
function calculatePolygonArea(points: Point[]): number {
  let area = 0;
  const n = points.length;

  for (let i = 0; i < n; i++) {
    const j = (i + 1) % n;
    area += points[i].x * points[j].y;
    area -= points[j].x * points[i].y;
  }

  return Math.abs(area / 2);
}

File: src-tauri/src/utils/mask_to_polygon.rs
/// Moore-Neighbor tracing algorithm for extracting contours from binary masks.

#[derive(Debug, Clone, Copy)]
pub struct Point {
    pub x: i32,
    pub y: i32,
}

/// Extract contour from a binary mask using Moore-Neighbor tracing.
/// `mask` is a flat array of booleans (row-major), true = foreground.
pub fn mask_to_polygon(mask: &[bool], width: usize, height: usize) -> Vec<Point> {
    // Find starting point (first true pixel from top-left)
    let mut start_x: Option<usize> = None;
    let mut start_y: Option<usize> = None;

    'outer: for y in 0..height {
        for x in 0..width {
            if mask[y * width + x] {
                start_x = Some(x);
                start_y = Some(y);
                break 'outer;
            }
        }
    }

    let (start_x, start_y) = match (start_x, start_y) {
        (Some(sx), Some(sy)) => (sx, sy),
        _ => return Vec::new(),
    };

    moore_neighbor_trace(mask, start_x, start_y, width, height)
}

/// Extract multiple contours from a mask (handles disconnected regions).
pub fn mask_to_polygons(mask: &[bool], width: usize, height: usize, min_area: f64) -> Vec<Vec<Point>> {
    let mut visited = vec![false; width * height];
    let mut polygons = Vec::new();

    for y in 0..height {
        for x in 0..width {
            let idx = y * width + x;
            if mask[idx] && !visited[idx] {
                let contour = moore_neighbor_trace(mask, x, y, width, height);

                // Mark visited pixels
                for p in &contour {
                    let px = p.x as usize;
                    let py = p.y as usize;
                    if px < width && py < height {
                        visited[py * width + px] = true;
                    }
                }

                // Filter by minimum area
                if contour.len() >= 3 {
                    let points: Vec<(f64, f64)> = contour.iter().map(|p| (p.x as f64, p.y as f64)).collect();
                    let area = super::converters::polygon_area(&points);
                    if area >= min_area {
                        polygons.push(contour);
                    }
                }
            }
        }
    }

    polygons
}

fn moore_neighbor_trace(mask: &[bool], start_x: usize, start_y: usize, width: usize, height: usize) -> Vec<Point> {
    let mut contour = Vec::new();

    // 8-connected neighbors (clockwise from N)
    // 0=N, 1=NE, 2=E, 3=SE, 4=S, 5=SW, 6=W, 7=NW
    let dx: [i32; 8] = [0, 1, 1, 1, 0, -1, -1, -1];
    let dy: [i32; 8] = [-1, -1, 0, 1, 1, 1, 0, -1];

    let mut x = start_x as i32;
    let mut y = start_y as i32;
    let mut direction: usize = 7; // Start looking from NW

    let max_iterations = width * height * 2;
    let mut iterations = 0;

    loop {
        contour.push(Point { x, y });

        let mut found = false;
        for i in 0..8 {
            let check_dir = (direction + i) % 8;
            let nx = x + dx[check_dir];
            let ny = y + dy[check_dir];

            if nx >= 0 && (nx as usize) < width && ny >= 0 && (ny as usize) < height {
                if mask[ny as usize * width + nx as usize] {
                    x = nx;
                    y = ny;
                    direction = (check_dir + 6) % 8; // backtrack 2
                    found = true;
                    break;
                }
            }
        }

        if !found {
            break;
        }

        iterations += 1;
        if iterations > max_iterations {
            break;
        }

        if x == start_x as i32 && y == start_y as i32 && contour.len() >= 2 {
            break;
        }
    }

    // Remove duplicate start/end point
    if contour.len() > 1 {
        let last = contour.len() - 1;
        if contour[0].x == contour[last].x && contour[0].y == contour[last].y {
            contour.pop();
        }
    }

    contour
}
